### ISSUE TITLE
Unable to type vertical bar

### ISSUE THREAD

**Issue by vinz243** [PowerShell/Win32-OpenSSH#902]
**"OpenSSH for Windows" version**

`0.0.21.0`
 
**Server OperatingSystem**

`Debian 9.1`

**Client OperatingSystem**

Windows 10 1703 15063.632

**What is failing**

I can't type a vertical bar (`|`) in the SSH terminal. Even using KeePass auto-type it doesn't work. However, in powershell or cmd, I can normally type the vbar. The user password on my server does has the `|` character in it, so I  am able to login (since it works in powershell), but then I cant `sudo` since it wont type the vbar. In putty, I am able to type the `|` character correctly.

**Expected output**

Typing `|` results in `|`

**Actual output**

Typing `|` results in nothing

**Comment by bagajjal** [comment]
It works for me on windows 10. 
Here is the screen shot,
![image](https://user-images.githubusercontent.com/23668037/31192835-2cf2fd9a-a8f7-11e7-8d57-6b0cc9b572b7.png)

**Comment by DHowett-MSFT** [comment]
`\|` is `VK_OEM_5`.

```cpp
#define VK_OEM_5          0xDC  //  '\|' for US
```

If @vinz243 is not using a US keyboard layout, `|` may not be on `VK_OEM_5`.

**Comment by vinz243** [comment]
True enough, I'm using azerty keyboard

**Comment by bagajjal** [comment]
@vinz243 - Not sure why you are facing this issue.. we don't have anything specific to VK_OEM_5.
As per code we are sending "|" to ssh server.
Can you try the keybased authentication so that you can login from the win32-openssh client. 
After you login
1) execute "xxd" command
2) Type |
3) press Ctrl+D

Please paste the screenshot.

fyi, it should look like this,
![image](https://user-images.githubusercontent.com/23668037/31417673-f950809a-ade5-11e7-8e12-0a65ab5e7385.png)

**Comment by leppakorpi** [comment]
I have same issue on Finnish keyboard with any character that requires right AltGr key.

With 0.0.20.0 everything work and xxd reports "|" as ^@| or 007c
With 0.0.21.0 I can't type these characters nor paste them from the clipboard, xxd shows ^[| or 1b7c

When using cmd.exe or powershell.exe I can change windows keyboard layout any time and it affects ssh session instantly, us keyboard layout works and fi keyboard layout doesn't.

With Visual Studio Code integrated console things get stickier. If the terminal is started with us keyboard, it keeps working after change to Finnish keyboard. If the terminal is started with fi keyboard, changing it later to 'us' won't help. So the work around for this bug is: 1) change keyboard to US layout 2) start new terminal in Visual studio code 3) change keyboard back to FI layout 4) start ssh session in terminal and Finnish keyboard works as expected.

**Comment by vinz243** [comment]
@bagajjal here is the output:
```
$ xxd
^[|00000000: 1b7c                                     .|
```

Please note that I didn't have to use key based authentication, since I can type `| ` just fine when typing the password. Also, I had to press <kbd>Ctrl</kbd> + <kbd>D</kbd> twice

**Comment by MathiasMagnus** [comment]
@bagajjal Using Hungarian keyboard I have the same problem using v0.0.22 and most of the scripting and programming special chars on AltGr: `$;#&@<>[]{}|\`

This is the console output for all of them at once:
```
mnagy@MATTY-Z50-75:~$ xxd
^[$^[;^[#^[&^[@^[<^[>^[[^[]^[{^[}^[|^[\00000000: 1b24 1b3b 1b23 1b26 1b40 1b3c 1b3e 1b5b  .$.;.#.&.@.<.>.[
00000010: 1b5d 1b7b 1b7d 1b7c 1b5c                 .].{.}.|.\
```

**Comment by leppakorpi** [comment]
I think this issue was introduced by the commit https://github.com/PowerShell/openssh-portable/commit/8747626b6aa845718c51ab130692338647c187a4
contrib/win32/win32compat/tncon.c line 712

As a quick fix I would disable '\x1b' for right alt, but I guess someone Americans just can't survive with only one meta key.

**Comment by bagajjal** [comment]
@leppakorpi, @MathiasMagnus, @vinz243 -
Could you guys please verify if the attached private ssh.exe binary fixes your issue?
[ssh.zip](https://github.com/PowerShell/Win32-OpenSSH/files/1400315/ssh.zip)

**Comment by MathiasMagnus** [comment]
Indeed it did. Thanks. Hoping the fix get incorporated into the next release.

**Comment by bagajjal** [comment]
Thanks for confirming. It will be in the next release.

**Comment by fle108** [comment]
Thanks for the exe, it solved my problem too (was unable to type @ key) but i can't type other specialized keys from my french keyboard, typicaly keys with accent : éèù .
Do i need to create my own issue or not ?

**Comment by bagajjal** [comment]
@Tb0b - I have created the new issue #925.
FYI, as a workaround you can use putty.

**Comment by bagajjal** [comment]
@Tb0b - please try the attached private binary and let me know if it works for you.
[ssh-unicode.zip](https://github.com/PowerShell/Win32-OpenSSH/files/1409027/ssh-unicode.zip)

Here is the snapshot (with the private binary attached),
![image](https://user-images.githubusercontent.com/23668037/31917435-752aa242-b80c-11e7-934f-8d68618d6ba6.png)

**Comment by fle108** [comment]
Thanks for the .exe, it works for  éùè but now  @| and other alt-gr keys doesn’t work ☹

--
Blasta19

De : bagajjal
Envoyé le :mardi 24 octobre 2017 01:09
À : PowerShell/Win32-OpenSSH
Cc : Tb0b; Mention
Objet :Re: [PowerShell/Win32-OpenSSH] Unable to type vertical bar (#902)

@Tb0b - please try the attached private binary and let me know if it works for you.
ssh-unicode.zip
Here is the snapshot (with the private binary attached),

**Comment by bagajjal** [comment]
@Tb0b - Yes, the previous binary has only the unicode fix... 
The attached private ssh binary now has both the unicode and the altgr related fix.
[ssh-unicode-altgr.zip](https://github.com/PowerShell/Win32-OpenSSH/files/1412021/ssh-unicode-altgr.zip)

**Comment by MathiasMagnus** [comment]
Yes, it would be nice if all the fixes could make it into the next release. Hungarian layout here, none of the characters that require AltGr work:

~<>[]#&@{};$\|

Just about everything used for programming.

**Comment by bagajjal** [comment]
@MathiasMagnus - We forgot to merge the PR (code fix) to the latest release.. Now its merged so the fix will be available in the next release (Dec-Mid, 0.0.25.0)
